awx/api/serializers.py

```diff
@@ -2201,6 +2201,89 @@ def create(self, validated_data):
         return return_data
 
 
+class BulkHostDeleteSerializer(serializers.Serializer):
+    hosts = serializers.ListField(
+        allow_empty=False,
+        max_length=250,
+        write_only=True,
+        help_text=_('List of hosts ids to be deleted, e.g. [105, 130, 131, 200]'),
+    )
+
+    class Meta:
+        model = Host
+        fields = 'hosts'
+        read_only_fields = ()
+
+    def validate(self, attrs):
+        request = self.context.get('request', None)
+
+        # Validating the number of hosts to be deleted
+        if len(attrs['hosts']) > settings.BULK_HOST_MAX_DELETE:
+            raise serializers.ValidationError(
+                {
+                    "ERROR": 'Number of hosts exceeds system setting BULK_HOST_MAX_DELETE',
+                    "BULK_HOST_MAX_DELETE": settings.BULK_HOST_MAX_DELETE,
+                    "Hosts_count": len(attrs['hosts']),
+                }
+            )
+
+        # Getting list of all host objects, filtered by the list of the hosts to delete
+        attrs['in_inv'] = Host.objects.get_queryset().filter(pk__in=attrs['hosts'])
+
+        # Converting the queryset data in a dict. to reduce the number of queries when
+        # manipulating the data
+        hosts_data = attrs['in_inv'].values()
+
+        if len(attrs['in_inv']) == 0:
+            raise serializers.ValidationError({"ERROR": "There are no hosts to delete"})
+
+        if len(attrs['in_inv']) < len(attrs['hosts']):
+            hosts_exists = [host['id'] for host in hosts_data]
+            raise serializers.ValidationError({"ERROR": "Not all hosts are exists", "hosts": list(set(attrs['hosts']).difference(hosts_exists))})
+
+        # Getting all inventories that the hosts can be in
+        inv_list = list(set([host['inventory_id'] for host in hosts_data]))
+
+        errors = {"ERRORS": "Cannot delete host(s) from inventory", "Inventory": dict()}
+        for inv in Inventory.objects.get_queryset().filter(pk__in=inv_list):
+            if inv.kind != '':
+                errors["Inventory"][inv.id] = {
+                    "Inventory_name": inv.name,
+                    "reason": "Hosts can only be deleted from manual inventories (not smart or constructed types).",
+                }
```

Comment:
we should check this after we check permissions, otherwise we are letting users check existence/type of inventories before we know they have permission to them

Reply:
done